### PR TITLE
Fixed #19580 -- Unified behavior of reverse foreign key and many-to-many relations for unsaved instances.

### PR DESCRIPTION
**Pull request by raydeal**
The fix changes behaviour of reverse FK to be the same as M2M.

### CODE REVIEW

File: tests/many_to_one_null/tests.py [hunk 1]
```diff
@@ -138,5 +138,6 @@ def test_related_null_to_field(self):
         c1 = Car.objects.create()
         d1 = Driver.objects.create()
         self.assertIs(d1.car, None)
-        with self.assertNumQueries(0):
-            self.assertEqual(list(c1.drivers.all()), [])
+        msg = '"<Car: Car object (1)>" needs to have a value for field "make" before this reverse many-to-one relationship can be used'
+        with self.assertRaisesMessage(ValueError, msg):
+            c1.drivers.all()
```
Comment (felixxm) [inline tests/many_to_one_null/tests.py:143]: Why this error is raised? This should return an empty list without raising an exception.

File: django/db/models/fields/related_descriptors.py [hunk 2]
```diff
@@ -577,6 +577,25 @@ def __init__(self, instance):
...
+            self.related_val = self._get_related_val(instance)
+
+            # Even if this relation is not to pk, we require still pk value.
+            # The wish is that the instance has been already saved to DB,
+            # although having a pk value isn't a guarantee of that.
+            if instance.pk is None:
+                raise ValueError("%r instance needs to have a primary key value before "
+                                 "a reverse many-to-one relationship can be used." %
+                                 instance.__class__.__name__)
+
```
Comment (felixxm) [inline django/db/models/fields/related_descriptors.py:590]: This is untested.

**Review by felixxm — COMMENTED** [review]
@raydeal Thanks for this patch :+1: However, as far as I'm aware the previous patch #13784 was closer to the expected behavior.

File: docs/spelling_wordlist [hunk 3]
```diff
@@ -48,6 +48,7 @@ Bcc
 BCC'ed
 bcrypt
 Beaven
+behaviour
```
[the remark this answers is not in the document]
Comment (raydeal) [reply docs/spelling_wordlist:51]: I have copied it from previous patch. When should a word  be added to this list?

[the remark this answers is not in the document]
Comment (smithdc1) [reply docs/spelling_wordlist:51]: The docs are written in en-us where behavior is the correct spelling. Likely the previous author was from elsewhere in the world where behaviour is preferred.

File: tests/many_to_one_null/tests.py (hunk 1 above)
Comment (raydeal) [reply tests/many_to_one_null/tests.py:143]: I have put comment in ticket about that.

File: docs/releases/4.1.txt [hunk 4]
```diff
@@ -323,6 +323,23 @@ this difference. In Django 4.0 and earlier,
...
+
+Now it has changed for foreign key:
+
+* For a foreign key relation, trying to add/remove/clear objects when the
+  relation field is null is now an error (no change for m2m).
+
+* For a foreign key relation, trying to use a relation of any kind
+  when the object isn't saved is now an error
+  (same behaviour as m2m - no change for m2m).
+
```
Comment (felixxm) [inline docs/releases/4.1.txt:342]: I would simplify this, e.g.
```
In order to unify behavior with many-to-many relations for unsaved model
instances, a reverse foreign key now raises ``ValueError`` when calling
:class:`related managers<django.db.models.fields.related.RelatedManager>` for
unsaved objects.
```

File: docs/topics/db/models.txt [hunk 5]
```diff
@@ -363,6 +363,63 @@ want. For example::
...
+        empty_poll.choice_set.clear()
+
+    * For a foreign key relation, trying to use a relation of any kind when the
+      object isn't saved is now an error (no change for m2m).
+      For example invoking::
+
+        Poll(question="How?").choice_set
+
+      now raises a ``ValueError``.
+      Previously it would return a ``RelatedManager``.
```
Comment (felixxm) [inline docs/topics/db/models.txt:421]: We don't need to be so descriptive, this PR unifies behavior of an edge case for unsaved model instances, so a release should be enough. I'd chop a `versionchanged` annotation.

File: django/db/models/fields/related_descriptors.py [hunk 6]
```diff
@@ -684,18 +701,21 @@ def check_and_update_obj(obj):
...
 
         def get_or_create(self, **kwargs):
+            self._check_fk_val()
             kwargs[self.field.name] = self.instance
             db = router.db_for_write(self.model, instance=self.instance)
             return super(RelatedManager, self.db_manager(db)).get_or_create(**kwargs)
         get_or_create.alters_data = True
 
         def update_or_create(self, **kwargs):
+            self._check_fk_val()
```
Comment (felixxm) [inline django/db/models/fields/related_descriptors.py:718]: This is untested.

**Review by felixxm — COMMENTED** [review]
@raydeal Thanks for updates :+1: I left more comments. Please rebase your patch and reformat the code with `black`.

File: django/db/models/fields/related_descriptors.py (hunk 6 above)
Comment (raydeal) [reply django/db/models/fields/related_descriptors.py:790]: test added

**Review by felixxm — COMMENTED** [review]
@raydeal Thanks :+1: Welcome aboard :boat:

**Comment by raydeal** [comment]
@felixxm thank you also for the constructive comments that help me to get more knowledge about django code and contribution.
@paveldroo thanks for former investigation and preparing a background of code.

**Pull request merged**